RelationalRecord/src/Relations/Role.hs

```diff
@@ -0,0 +1,111 @@
+{-# LANGUAGE TemplateHaskell, MultiParamTypeClasses, FlexibleInstances #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module  Relations.Role where
+
+import  Types.Role          as Role
+import  Types.User          as User
+import  Types.UsersRoles    as UsersRoles
+import  Types.DB
+import  Relations.DB
+
+import  Database.Relational.Query
+import  Database.HDBC.Query.TH      (makeRecordPersistableDefault)
+
+
+-- SELECTS
+
+allRoles :: Relation () Roles
+allRoles = Role.roles
+
+allRoleAssignments :: Relation () UsersRoles
+allRoleAssignments = UsersRoles.usersRoles
+
+getRole :: Relation PKey Roles
+getRole = relation' . placeholder $ \rolId -> do
+    a       <- query roles
+    wheres  $ a ! Role.id' .=. rolId
+    return  a
+
+-- given a user, get all his/her roles (inner join)
+getRoles :: Relation Users Roles
```

Comment:
Can you please give a usage example of this function? Is this just a DSL/AST representation of an inner-join between `roles` and `users` via the `user_roles` join-through table, with an unapplied `user` argument? What is the type of the unapplied `user` argument? Is it the `PKey` of the `User` table or the actual `User` record? How can you tell which arguments are unapplied from the function signature?

When this query is executed, what does it evaluate to? A list of `Roles`?

How would one write the following -- get **all** users along with each user's role(s), where the user has been created after a given timestamp AND the user's email ID belongs to a given domain (say, gmail.com or vacationlabs.com)?

Can I compose this `Relation` with another `wheres` clause specified outside this function? Can I compose this with another JOIN, say with `tenants` table?